Shrink should not touch max_retries

henningandersen opened the pull request:
Shrink would set `max_retries=1` in order to avoid retrying. This
however sticks to the shrunk index afterwards, causing issues when a
shard copy later fails to allocate just once.

Avoiding a retry of a shrink makes sense since there is no new node
to allocate to and a retry will likely fail again. However, the downside of
having `max_retries=1` afterwards outweigh the benefit of not retrying
the failed shrink a few times. This change ensures shrink no longer
sets `max_retries` and also makes all resize operations (shrink, clone,
split) leave the setting at default value rather than copy it from source.

henningandersen committed:
Shrink should not touch max_retries

Shrink would set `max_retries=1` in order to avoid retrying. This
however sticks to the shrunk index afterwards, causing issues when a
shard copy later fails to allocate just once. While there is no new node
to allocate to and a retry will likely fail again, the downside of
having `max_retries=1` afterwards outweigh the benefit of not retrying
the failed shrink a few times. This change ensures shrink no longer
sets `max_retries`.

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Allocation)

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

jasontedor commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -766,10 +766,7 @@ static void prepareResizeIndexSettings(
-                .put(initialRecoveryIdFilter, Strings.arrayToCommaDelimitedString(nodesToAllocateOn.toArray()))
-                // we only try once and then give up with a shrink index
-                .put("index.allocation.max_retries", 1);
+            indexSettingsBuilder.put(initialRecoveryIdFilter, Strings.arrayToCommaDelimitedString(nodesToAllocateOn.toArray()));
```
Note that we will now inherit the value of this setting if it's present already on the source index.

jasontedor commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexServiceTests.java:
```diff
@@ -267,7 +268,7 @@ public void testPrepareResizeIndexSettings() {
                             equalTo("keyword"));
                     assertThat(settings.get("index.routing.allocation.initial_recovery._id"), equalTo("node1"));
-                    assertThat(settings.get("index.allocation.max_retries"), equalTo("1"));
+                    assertThat(settings.get("index.allocation.max_retries"), nullValue());
```
Let's say we keep the production implementation as-is.

Do we need to continue to assert on this at all? It feels odd to single this setting as unset amongst all the other settings, it's only an accident of history that we're making an assertion here. Also, we would inherit this setting if it was present on an existing index (and an open question whether or not we should do that, I think we should, but I'm curious to hear your perspective) so this assertion can be misleading then (even though it's true in this test, because the index wasn't set up with this setting set).

If we do change the production implementation, to override any existing value for `index.allocation.max_retries`, then the assertion makes sense, but we should also consider covering the case where the setting is present on the source index.

jasontedor replied:
I edited the above comment to add a key missing not. To be clear, I think that we should *not* inherit any existing value for this setting, but instead override it. We have a built-in mechanism for this, to set not copyable on resize.

jasontedor replied:
And I don't think we need to worry about whether or not this a breaking change. Today we are universally overriding the value of the setting to one (whether it's set or unset). With the change as implemented, we would now inherit the value. If we change the production implementation to never copy the value for this setting, that wouldn't be breaking, since we weren't preserving it already today.

henningandersen replied:
I did think about removing that assertion, but also felt like I then did not really test the change. Ideally the test would test that only expected properties are present instead and I could make that change instead, as well as ensure it now copies it from source if set (depending on the copy-setting discussion).

About copying the setting, I think we should. I believe that is the intended direction after #30255? If this setting had been set explicitly on the source, I think it makes sense to carry it through shrinking, at least in an ILM context. Additionally, this would AFAICS be the only setting that would not be copied if we add this. I am curious to know why you are considering to not copy this specific setting?

jasontedor replied:
I think it's a setting of dubious value, and I'm not sure why we even have it (at the index level). We are already overriding it in resizes today, and we haven't seen complaints about that. We don't replicate it in cross-cluster replication.

jasontedor reviewed (commented):
Thanks @henningandersen I agree with your assessment, I left some comments though.

henningandersen committed:
Resize: no longer copy max_retries

If max_retries was set on source, it is unlikely to be wanted on target
too, instead the new index will rely on the default.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into enhance_shrink_not_touch_max_retries

henningandersen commented:
@elasticmachine run elasticsearch-ci/packaging-sample

jasontedor reviewed (approved):
LGTM.

henningandersen commented:
Thanks @jasontedor
